src/lib/TopLevel.hs

```diff
@@ -241,39 +226,20 @@ requestArrays backend vs = case backend of
     callPipeServer (psPop server) vs'
   _ -> error "Not implemented"
 
-substArrayLiterals :: BackendEngine -> Atom -> IO Atom
+substArrayLiterals :: (HasVars a, HasType a) => BackendEngine -> a -> IO a
 substArrayLiterals backend x = do
-  x' <- substAtomTypes x
-  let vs = arrayVars x'
-  arrays <- requestArrays backend vs
-  let arrayTypes = fmap varAnn vs
-  let arrayAtoms = fmap (Con . uncurry ArrayLit) $ zip arrayTypes arrays
-  return $ subst (newLEnv vs arrayAtoms, mempty) x'
-  where
-    -- Substitutes all arrays appearing in the type annotations inside x
-    substAtomTypes :: Atom -> IO Atom
-    substAtomTypes a = case a of
-      Var v -> Var <$> traverse substInType v
-      -- No substitutions under lambdas. This might cause some issues if we wanted
-      -- to print their body, but it's questionable whether we even want to do that.
-      Con c -> Con <$> traverseExpr c substInType substAtomTypes return
-      _     -> error $ "Unexpected atom: " ++ pprint x
+  x' <- substArrayLiterals' backend (arrayVars (getType x)) x
```

Comment:
Might be worth adding a comment mentioning the short correctness proof we've arrived at yesterday